c2rust-analyze/src/known_fn.rs

```diff
@@ -602,28 +622,823 @@ pub const fn all_known_fns() -> &'static [KnownFn] {
+                maxevents: c_int,
+                timeout: c_int,
+            ) -> c_int;
+
+            fn execv(
+                prog: *const c_char: [READ | OFFSET_ADD | NON_NULL],
+                argv: *const *const c_char: [READ | OFFSET_ADD | NON_NULL, READ | OFFSET_ADD],
+            ) -> c_int;
+
+            fn execve(
+                prog: *const c_char: [READ | OFFSET_ADD | NON_NULL],
+                argv: *const *const c_char: [READ | OFFSET_ADD | NON_NULL, READ | OFFSET_ADD],
+                envp: *const *const c_char: [READ | OFFSET_ADD | NON_NULL, READ | OFFSET_ADD],
+            ) -> c_int;
+
+            fn exit(
+                status: c_int,
+            ) -> !;
+
+            #[cfg(target_os = "linux")]
+            fn explicit_bzero(
+                s: *mut c_void: [WRITE | OFFSET_ADD | NON_NULL],
+                len: size_t,
+            ) -> ();
+
+            fn fchdir(
+                dirfd: c_int,
+            ) -> c_int;
+
+            // fn fcntl(
+            //     fd: c_int,
+            //     cmd: c_int,
+                   // TODO(kkysen) varargs
+            //     ...
+            // ) -> c_int,
+
+            fn fflush(
+                file: *mut FILE: [READ | WRITE | NON_NULL],
+            ) -> c_int;
+
+            fn fork() -> pid_t;
+
+            // fn fprintf(
+            //     stream: *mut FILE: [READ | WRITE | NON_NULL],
+            //     format: *const c_char: [READ | OFFSET_ADD | NON_NULL],
+                   // TODO(kkysen) varargs, printf-style
+            //     ...
+            // ) -> c_int;
+
+            fn fputs(
+                s: *const c_char: [READ | OFFSET_ADD | NON_NULL],
+                stream: *mut FILE: [READ | WRITE | NON_NULL],
+            ) -> c_int;
+
+            fn free(
+                p: *mut c_void: [FREE],
+            ) -> ();
+
+            fn freeaddrinfo(
+                // TODO(kkysen) Should this be `FREE`, as it could be manually `free`d,
+                // but is not meant to be `free`d directly?
+                res: *mut addrinfo: [READ | WRITE],
+            ) -> ();
+
+            fn fstat(
+                fildes: c_int,
+                buf: *mut stat: [WRITE | NON_NULL],
+            ) -> c_int;
+
+            fn ftruncate(
+                fd: c_int,
+                length: off_t,
+            ) -> c_int;
+
+            fn gai_strerror(
+                errcode: c_int,
+            ) -> *const c_char: [READ | OFFSET_ADD | NON_NULL];
+
+            fn getaddrinfo(
+                node: *const c_char: [READ | OFFSET_ADD],
+                service: *const c_char: [READ | OFFSET_ADD],
+                hints: *const addrinfo: [READ],
+                // TODO(kkysen) Should the 2nd ptr be `FREE`, as it could be manually `free`d,
+                // but is not meant to be `free`d directly, but by `freeaddrinfo`?.
+                res: *mut *mut addrinfo: [WRITE | NON_NULL, WRITE | NON_NULL],
```

Comment:
Good question.  I sort of expect that we'll end up tying `FREE` to `Box<T>`/`Rc<T>`/etc, which don't (yet) support custom deallocators like `freeaddrinfo`.  But maybe we could handle this with an additional permission or flag that translates into a custom "linear (non-affine) box" type that must be explicitly freed with a special function (you can't just let it go out of scope).

Reply:
That sounds like a potentially good way.  I'll add some of that to the comment and leave it at that for now if that sounds good.

Reply:
I'll open an issue for this, too.  There are also ones like `glob`/`globfree` and `mmap`/`munmap`.  That said, even if we don't track them correctly, memory leaks are safe, and we'll be doing the same memory management as was in C.  It'd be similar to any other resource as well, like `open`, etc. and `close`.

Reply:
* Opened #1007.